### ISSUE TITLE
Use Chromecast for keyboard inputs, rather than ADB

### ISSUE THREAD

**Issue by ROBOT0-VT** [Nerwyn/universal-remote-card#89]
**Is your feature request related to a problem? Please describe.**
Can't use keyboard without enabling ADB (which increases the device's security risk)

**Describe the solution you'd like**
The keyboard functionality should use the cast or remote integrations (or another method), rather than ADB; in a similar way to the google TV app

**Describe alternatives you've considered**
N/A

**Additional context**
I've had a look at the following similar issues:
- #55: No mention of ADB
- #82: Mentions ADB, but from a performance perspective, not a security one
- #4: Original issue for keyboard support, closed
- [Issue in underlying python library](https://github.com/tronikos/androidtvremote2/issues/13)

I believe this issue should be tracked, and closed once a fix is possible

**Comment by Nerwyn** [comment]
This is a dead end until someone reverse engineers the Android TV Remote API keyboard commands.

**Comment by ROBOT0-VT** [comment]
Agreed, but it should still be noted

**Comment by astrokin** [comment]
I’ve finally located the option to send text, and I’m hoping to submit a PR soon. If anyone needs this feature urgently, please feel free to reach out to me directly.

**Comment by Nerwyn** [comment]
Via the Android TV Remote API? That's amazing! Keep us posted! I'll add support for it in this card once it's available as a Home Assistant action.

**Comment by astrokin** [comment]
@Nerwyn yes. without using ADB. the issue with protofiles using everywhere. they are incorrect.

**Comment by Nerwyn** [comment]
Do you know if there's a way to trigger a global search without ADB? I was thinking about deprecating `keyboard_id` sicne it's only really needed for Android TV with the ADB keyboard, but I may need to keep it if there isn't a way to do so without ADB. Specifically this ADB command:

```
am start -a "android.search.action.GLOBAL_SEARCH" --es query "${text}"
```

**Comment by astrokin** [comment]
the only thing that worked for me was `KEYCODE_SEARCH = 84;`  but on my TV it trigger only voice input, which I did not implemented yet

**Comment by astrokin** [comment]
feel free to find text input implementation without ADB for Android TV for [python](https://github.com/tronikos/androidtvremote2/pull/25) and [js](https://github.com/louis49/androidtv-remote/pull/46)

**Comment by felipecrs** [comment]
@astrokin you are amazing. I guess the next step to move this forward is adding support for it in Home Assistant. Perhaps through some special prefix like:

```yaml
action: remote.send_command
data:
  command: text://Some User Written Text Here Yada Yada
```

**Comment by Nerwyn** [comment]
@felipecrs the implementation of Android TV Remote keyboard has been in debate/limbo for a few months now 

https://github.com/home-assistant/core/pull/139033

It looks like the HA maintainers are pushing for it to use the Android TV Remote `config_entry` instead.

**Comment by tronikos** [comment]
@Nerwyn with 2025.6 you should be able to send text with:

```yaml
action: remote.send_command
data:
  command: text:hello world
target:
  entity_id: remote.my_tv
```

**Comment by Nerwyn** [comment]
EXCELLENT.  I'll start working on the next minor version to implement this in the next few days. I do see that https://github.com/home-assistant/core/pull/139033 and https://github.com/home-assistant/core/pull/142438 never closed. Did you have another PR that implemented this? I'm glad you got the `remote.send_command` syntax approved, I do not understand why the config entry method was being suggested (I would've just had users provide the remote entity and done the config entry lookup using the hass object so it wouldn't have be a problem here, but it would've been confusing for people not using this card).

You rock @tronikos! I only switched from home theater PC to a Chromecast with Google TV and started this project (and Home Assistant frontend open source projects in general) because you implemented the Android TV Remote API into Home Assistant. Now I spend half my free time working on this stuff!

**Comment by MahdiBM** [comment]
@Nerwyn thank you. I tried this with 2025.6.0 hass core and the alpha you released and the good news is that the other integrations have started working in the keyboard. Like Sony Bravia or Google Cast.

The interesting thing is that they still exhibit the same behavior. Right now the only consistent keys seems to be dot and colon (.,) and if you use any other keys the result might not end up on TV. Numbers work meh. Characters just never end up on TV.

This suggests that there is something unrelated to the integrations that filters/blocks characters 🤔

**Comment by felipecrs** [comment]
> @Nerwyn @felipecrs I'm grad to help you!

I was only helped as a user btw, @Nerwyn should take all the credit. I'm very thankful for you both, though.

**Comment by Nerwyn** [comment]
I've put out the 4.6.0 beta/release candidate if any of y'all want to spot test the new keyboard and search methods. The search action now calls the `SEARCH` key followed by some dpad navigation to get to search keyboard input.

```yaml
action: remote.send_command
data:
  delay_secs: 1
  command:
    - SEARCH
    - DPAD_LEFT
    - DPAD_LEFT
    - DPAD_CENTER
target:
  entity_id: remote.lounge_google_tv
```

If this doesn't work well for you, you can still use the old ADB method by creation a custom element with a search action and platform Fire TV (which uses ADB).

**Comment by Nerwyn** [comment]
Added in [4.6.0](https://github.com/Nerwyn/universal-remote-card/releases/tag/4.6.0).
